Make sure build-essential is installed

From system log:


[   58.949734] cloud-init[1212]: Granting passwordless sudo to JupyterHub admins...
[   58.960084] cloud-init[1212]: Setting up user environment...
[   58.964512] cloud-init[1212]: Downloading & setting up user environment...
[  153.133719] cloud-init[1212]: Traceback (most recent call last):
[  153.152077] cloud-init[1212]:   File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
[  153.164082] cloud-init[1212]:     "__main__", mod_spec)
[  153.180083] cloud-init[1212]:   File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
[  153.196282] cloud-init[1212]:     exec(code, run_globals)
[  153.207880] cloud-init[1212]:   File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/installer.py", line 453, in <module>
[  153.220003] cloud-init[1212]:     main()
[  153.227572] cloud-init[1212]:   File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/installer.py", line 436, in main
[  153.239993] cloud-init[1212]:     ensure_user_environment(args.user_requirements_txt_url)
ci-info: +++++Authorized keys from /home/ubuntu/.ssh/authorized_keys for user ubuntu++++++
[  153.257758] cloud-init[1212]:   File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/installer.py", line 255, in ensure_user_environmentci-info: +---------+-------------------------------------------------+---------+---------+
ci-info: | Keytype |                Fingerprint (md5)                | Options | Comment |

[  153.281450] cloud-init[1212]: ci-info: +---------+-------------------------------------------------+---------+---------+
    'ipywidgets==7.4.2'
[  153.294392] cloud-initci-info: | ssh-rsa | d9:a2:19:01:fe:6c:e7:9b:96:a3:de:68:d8:16:4c:52 |    -    | aws_key |
[1212]:   File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/conda.py", line 133, in ensure_pip_packagesci-info: +---------+-------------------------------------------------+---------+---------+

[  153.318441] cloud-init[1212]:     ] + packages, stderr=subprocess.STDOUT)
[  153.328373] cloud-init[1212]:   File "/usr/lib/python3.6/subprocess.py", line 336, in check_output
[  153.336818] cloud-init[1212]:     **kwargs).stdout
[  153.341928] cloud-init[1212]:   File "/usr/lib/python3.6/subprocess.py", line 418, in run
[  153.348801] cloud-init[1212]:     output=stdout, stderr=stderr)
[  153.360976] cloud-init[1212]: subprocess.CalledProcessError: Command '['/opt/tljh/user/bin/python', '-m', 'pip', 'install', '--no-cache-dir', 'jupyterhub==0.9.4', 'notebook==5.7.0', 'jupyterlab==0.35.3', 'nteract-on-jupyter==1.9.12', 'nbgitpuller==0.6.1', 'nbresuse==0.3.0', 'ipywidgets==7.4.2']' returned non-zero exit status 1.
[  153.378331] <14>Feb 27 15:30:12 ec2: 
<14>Feb 27 15:30:12 ec2: 
cloud-init[1212]: Cloud-init v. 18.3-9-g2e62cb8a-0ubuntu1~18.04.2 running 'modules:final' at Wed, 27 Feb 2019 15:28:00 +0000. Up 20.59 seconds.
[  153.465289] cloud-init[1212]: 2019-02-27 15:30:12,792 - util.py[WARNING]: Failed running /var/lib/cloud/instance/scripts/part-001 [1]
[  153.477232] cloud-init[1212]: 2019-02-27 15:30:12,797 - cc_scripts_user.py[WARNING]: Failed to run module scripts-user (scripts in /var/lib/cloud/instance/scripts)
[  153.496913] cloud-init[1212]: 2019-02-27 15:30:12,798 - util.py[WARNING]: Running module scripts-user (<module 'cloudinit.config.cc_scripts_user' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_scripts_user.py'>) failed
[  153.521355] cloud-init[1212]: Cloud-init v. 18.3-9-g2e62cb8a-0ubuntu1~18.04.2 finished at Wed, 27 Feb 2019 15:30:13 +0000. Datasource DataSourceEc2Local.  Up 153.46 seconds
[[0;1;31mFAILED[0m] Failed to start Execute cloud user/final scripts.
See 'systemctl status cloud-final.service' for details.
[[0;32m  OK  [0m] Reached target Cloud-init target.
Close

Running the failing command by hand reveal the real error:
```
    /usr/lib/gcc/x86_64-linux-gnu/7/include-fixed/limits.h:194:15: fatal error: limits.h: No such file or directory
     #include_next <limits.h>  /* recurse down to the real one */
```
The documentation should specify that the build-essential package needs to be installed.

--- a/tljh/installer.py
+++ b/tljh/installer.py
@@ -187,6 +187,18 @@ def ensure_jupyterhub_package(prefix):
     hub environment be installed with pip prevents accidental mixing of python
     and conda packages!
     """
+    # Install pycurl. JupyterHub prefers pycurl over SimpleHTTPClient automatically
+    # pycurl is generally more bugfree - see https://github.com/jupyterhub/the-littlest-jupyterhub/issues/289
+    # build-essential is also generally useful to everyone involved, and required for pycurl
+    apt.install_packages([
+        'libssl-dev',
+        'libcurl4-openssl-dev',
+        'build-essential'
+    ])
+    conda.ensure_pip_packages(prefix, [
+        'pycurl==7.43.*'
+    ])
+
     conda.ensure_pip_packages(prefix, [
         'jupyterhub==1.0.0',
         'jupyterhub-dummyauthenticator==0.3.1',
@@ -228,11 +240,6 @@ def ensure_user_environment(user_requirements_txt_file):
         logger.info('Downloading & setting up user environment...')
         with conda.download_miniconda_installer(miniconda_version, miniconda_installer_md5) as installer_path:
             conda.install_miniconda(installer_path, USER_ENV_PREFIX)
-
-    # nbresuse needs psutil, which requires gcc
-    apt.install_packages([
-        'gcc'
-    ])
 
     conda.ensure_conda_packages(USER_ENV_PREFIX, [
         # Conda's latest version is on conda much more so than on PyPI.

